### PR TITLE
feat(Vault): add exchange rate

### PR DESCRIPTION
**Pull request by PierrickGT**

### CODE REVIEW

File: test/unit/Vault/Undercollateralization.t.sol [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    uint256 _yield = 400_000e18;
+    _accrueYield(underlyingAsset, yieldVault, _yield);
+
+    // And increase yield fee balance
+    // vault.increaseYieldFeeBalance(_yield);
+
+    underlyingAsset.burn(address(yieldVault), 10_000_000e18);
+
+    assertEq(vault.isVaultCollateralized(), false);
+
```
Comment (asselstine) [inline test/unit/Vault/Undercollateralization.t.sol:98]: Maybe also assert here that there are zero yield fees?

Comment (PierrickGT) [reply test/unit/Vault/Undercollateralization.t.sol:98]: Yield fee is 400,000 and now that the vault is undercollateralized, if this fee has not been claimed yet, users can withdraw it.
They'll receive part of the yield proportionally to their deposit.

**Review by dylandesrosier — COMMENTED** [review]
How would the `Vault` become undercollatoralized without the yield also being at risk? 
It seems like the underlying `YieldVault` being malicious would be the main cause of this. In that case, we wouldn't be able to trust the `_yieldVault.maxWithdraw` => can't trust `_totalAssets` => can't trust `availableYieldBalance` => can't trust `availableYieldFeeBalance`. So using the available fees as a fallback allocation of tokens to re-collatoralize the users deposits likely won't work since those tokens will also be gone/shrunk/undercollatoralized.

It seems like another source of problems is distributing tokens sent directly to the `Vault` to the depositors. This transfer causes the exchange rate between assets and shares in the `Vault` to grow (more assets per share). 
Could we allocate tokens sent to the `Vault` to be liquidated rather than distributed to depositors? Then we can rely on the `yieldVault.maxWithdraw` and `_assetSupplyBalance` for the allocations of tokens. `yieldVault.maxWithdraw - _assetSupplyBalance` is the available yield (which can be further broken down for the amount to be liquidated and the portion for fees) and _assetSupplyBalance stays 1:1 with the amount supplied to the `YieldVault`, meaning the exchange rate for shares to assets stays 1:1.

File: test/unit/Vault/Deposit.t.sol [hunk 2]
```diff
@@ -126,6 +126,87 @@ contract VaultDepositTest is UnitBaseSetup {
...
+
+    uint256 _attackAmount = 1000e18;
+    uint256 _bobAmount = 1;
+    uint256 _bobTotalAmount = _attackAmount + _bobAmount;
+
+    underlyingAsset.mint(bob, _attackAmount + _bobAmount);
+
+    _deposit(underlyingAsset, vault, _bobAmount, bob);
+
+    // Bob deposits 1,000 underlying assets into the Vault in the hope of manipulating the exchange rate
```
[the remark this answers is not in the document]
Comment (PierrickGT) [reply test/unit/Vault/Deposit.t.sol]: Fixed: https://github.com/pooltogether/v5-vault/pull/7/commits/38a34672231160d020c6170b85cca6024684760f

File: src/Vault.sol [hunk 3]
```diff
@@ -201,26 +208,35 @@ contract Vault is ERC4626, ERC20Permit, ILiquidationSource, Ownable {
...
+
   /// @inheritdoc ILiquidationSource
   function availableBalanceOf(address _token) public view override returns (uint256) {
     return _availableBalanceOf(_token);
   }
 
   /**
-   * @notice Get the total available yield amount accrued by this vault.
+   * @notice Total available yield amount accrued by this vault.
    * @dev This amount includes the liquidable yield + yield fee amount.
```
[the remark this answers is not in the document]
Comment (PierrickGT) [reply src/Vault.sol]: Fixed: https://github.com/pooltogether/v5-vault/pull/7/commits/38a34672231160d020c6170b85cca6024684760f

**Comment by PierrickGT** [comment]
> How would the `Vault` become undercollatoralized without the yield also being at risk? It seems like the underlying `YieldVault` being malicious would be the main cause of this. In that case, we wouldn't be able to trust the `_yieldVault.maxWithdraw` => can't trust `_totalAssets` => can't trust `availableYieldBalance` => can't trust `availableYieldFeeBalance`. So using the available fees as a fallback allocation of tokens to re-collatoralize the users deposits likely won't work since those tokens will also be gone/shrunk/undercollatoralized.

Yes exactly but the loss could be even greater if we don't allocate the yield to users. So in a worst case scenario, I feel like it makes sense to allocate the yield to users.

> It seems like another source of problems is distributing tokens sent directly to the `Vault` to the depositors. This transfer causes the exchange rate between assets and shares in the `Vault` to grow (more assets per share). Could we allocate tokens sent to the `Vault` to be liquidated rather than distributed to depositors? Then we can rely on the `yieldVault.maxWithdraw` and `_assetSupplyBalance` for the allocations of tokens. `yieldVault.maxWithdraw - _assetSupplyBalance` is the available yield (which can be further broken down for the amount to be liquidated and the portion for fees) and _assetSupplyBalance stays 1:1 with the amount supplied to the `YieldVault`, meaning the exchange rate for shares to assets stays 1:1.

Currently, the exchange rate is calculated by using `totalSupply`, so the amount of underlying assets living in the vault does not impact the exchange rate.
https://github.com/pooltogether/v5-prize-vault/blob/38a34672231160d020c6170b85cca6024684760f/src/Vault.sol#L981
https://github.com/pooltogether/v5-prize-vault/blob/38a34672231160d020c6170b85cca6024684760f/src/Vault.sol#L701

The underlying assets sent to the Vault are currently allocated to be distributed since we use `totalAssets` to calculate `availableYieldBalance`.
So if there are any underlying assets living in the Vault, they are contributed.
https://github.com/pooltogether/v5-prize-vault/blob/38a34672231160d020c6170b85cca6024684760f/src/Vault.sol#L228
https://github.com/pooltogether/v5-prize-vault/blob/38a34672231160d020c6170b85cca6024684760f/src/Vault.sol#L693
https://github.com/pooltogether/v5-prize-vault/blob/38a34672231160d020c6170b85cca6024684760f/src/Vault.sol#L469

Also when depositing, if there are underlying assets living in the Vault, they are deposited into the YieldVault.
This way, user may need to deposit less underlying assets if there are already some in the Vault and inflation attacks are not possible since any underlying assets living in the Vault are sent to the YieldVault.
https://github.com/pooltogether/v5-prize-vault/blob/38a34672231160d020c6170b85cca6024684760f/src/Vault.sol#L818
https://github.com/pooltogether/v5-prize-vault/blob/38a34672231160d020c6170b85cca6024684760f/test/unit/Vault/Deposit.t.sol#L169

**Pull request merged**